x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Lookup.java

```diff
@@ -22,86 +23,62 @@
-    ) {
+    private final Expression tableName;
+    private final List<NamedExpression> matchFields;
+    // initialized during the analysis phase for output and validation
+    // afterward, it is converted into a Join (BinaryPlan) hence why here it is not a child
+    private final LocalRelation localRelation;
+    private List<Attribute> lazyOutput;
+
+    public Lookup(Source source, LogicalPlan child, Expression tableName, List<NamedExpression> matchFields) {
+        this(source, child, tableName, matchFields, null);
+    }
+
+    public Lookup(Source source, LogicalPlan child, Expression tableName, List<NamedExpression> matchFields, LocalRelation localRelation) {
         super(source, child);
         this.tableName = tableName;
         this.matchFields = matchFields;
-        this.matchValues = matchValues;
-        this.mergeValues = mergeValues;
+        this.localRelation = localRelation;
     }
 
-    public Attribute tableName() {
+    public Expression tableName() {
         return tableName;
     }
 
-    public List<Attribute> matchFields() {
+    public List<NamedExpression> matchFields() {
         return matchFields;
     }
 
-    public List<TableColumnAttribute> matchValues() {
-        return matchValues;
-    }
-
-    public List<TableColumnAttribute> mergeValues() {
-        return mergeValues;
+    public LocalRelation localRelation() {
+        return localRelation;
     }
 
     @Override
     public boolean expressionsResolved() {
-        if (matchValues == null || mergeValues == null) {
-            return false;
-        }
-        return tableName.resolved()
-            && Resolvables.resolved(matchFields)
-            && Resolvables.resolved(matchValues)
-            && Resolvables.resolved(mergeValues);
+        return tableName.resolved() && Resolvables.resolved(matchFields);
     }
 
     @Override
     public UnaryPlan replaceChild(LogicalPlan newChild) {
-        return new Lookup(source(), newChild, tableName, matchFields, matchValues, mergeValues);
+        return new Lookup(source(), newChild, tableName, matchFields);
     }
 
     @Override
     protected NodeInfo<? extends LogicalPlan> info() {
-        return NodeInfo.create(this, Lookup::new, child(), tableName, matchFields, matchValues, mergeValues);
+        return NodeInfo.create(this, Lookup::new, child(), tableName, matchFields);
     }
 
     @Override
     public List<Attribute> output() {
-        if (mergeValues == null) {
-            throw new IllegalStateException("unresolved");
-        }
-        if (mergeValues.isEmpty()) {
-            return child().output();
-        }
-        if (this.output == null) {
-            this.output = mergeOutputAttributes(mergeValues, child().output());
+        if (lazyOutput == null) {
+            List<? extends NamedExpression> rightSide = localRelation != null ? localRelation.output() : matchFields;
+            lazyOutput = mergeOutputAttributes(child().output(), rightSide);
```

Comment:
The local relation is not always available (no such table) - thus return the matching attribute instead of some fake/empty output list for the relationship.

Reply:
This'll fail late on, right?

Reply:
Which is kind of the point. We want these to generate "unresolved" errors.